core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala

```diff
@@ -72,8 +72,11 @@ private[spark] class StaticMemoryManager(
       numBytes: Long,
       evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = synchronized {
     val currentUnrollMemory = storageMemoryPool.memoryStore.currentUnrollMemory
-    val maxNumBytesToFree = math.max(0, maxMemoryToEvictForUnroll - currentUnrollMemory)
-    val numBytesToFree = math.min(numBytes, maxNumBytesToFree)
+    val freeMemory = storageMemoryPool.memoryFree
+    val maxNumBytesToFree = math.max(0,
+      maxMemoryToEvictForUnroll - currentUnrollMemory - freeMemory)
```

Comment:
E.g. Let's say we have 1000B of storage memory, 700B used by blocks and `maxUnrollMemory` is 400B. The max amount of memory we want to evict here is only 400 - 300 = 100B. If we did not subtract free memory from the equation, then we might end up evicting 400B worth of blocks, which is too much since unrolling will also use the existing free space.